docs/examples/crypto-sockets.md

```diff
@@ -0,0 +1,160 @@
+
+flow = Dataflow()
+flow.map(json.loads)
+flow.map(key_on_product)
+```
+
+Now for the exciting part. The code below is what we are using to: 
+1. Construct the orderbook as two dictionaries, one for asks, one for bids
+2. Assign a value to the ask and bid price.
+3. For each new order, update the order book and then update the bid and ask prices where required.
+4. Return bid and ask price, the respective volumes of the ask and the difference between the prices.
+
+So how does maintaining this `OrderBook` object work in a Dataflow. Below you will notice that we are using `stateful_map` as the operator for this aggregation step. `Stateful_map` will aggregate data based on a function (mapper), into an object that you define. The object must be defined via a builder, because it will create a new object via this builder for each new key received. The mapper must return the object so that it can be updated.
+
+
+```Python
+class OrderBook:
+    def __init__(self):
+        # if using Python < 3.7 need to use OrderedDict here
+        self.bids = {}
+        self.asks = {}
+
+    def update(self, data):
+        if self.bids == {}:
+            self.bids = {float(v[0]):float(v[1]) for v in data['bids']}
+            self.bid_price = next(iter(self.bids))
+        if self.asks == {}:
+            self.asks = {float(v[0]):float(v[1]) for v in data['asks']}
+            self.ask_price = next(iter(self.asks))
+        else:
+            for update in data['changes']:
+                price = float(update[1])
+                size = float(update[2])
+            if update[0] == 'sell':
+                # modify asks
+                if size == 0.0:
+                    try:
+                        del self.asks[price]
+                        if price <= self.ask_price:
+                            self.ask_price = sorted(self.asks.keys())[0]
+                    except KeyError:
+                        # don't need to add price with size zero
+                        pass
+                else:
+                    self.asks[price] = size
+                    if price < self.ask_price:
+                        self.ask_price = price
+            if update[0] == 'buy':
+                # modify bids
+                if size == 0.0:
+                    try:
+                        del self.bids[price]
+                        if price >= self.bid_price:
+                            self.bid_price = sorted(self.bids.keys())[-1]
+                    except KeyError:
+                        # don't need to add price with size zero
+                        pass
+                else:
+                    self.bids[price] = size
+                    if price > self.bid_price:
+                        self.bid_price = price
+        return self, (self.bid_price, self.bids[self.bid_price], self.ask_price, self.asks[self.ask_price], self.ask_price-self.bid_price)
+
+flow.stateful_map(lambda key: OrderBook(), OrderBook.update)
```

Comment:
Now that we have the basic recovery branch merged, this requires a `step_id` name as the first argument. You could use whatever: `flow.stateful_map("order_book", lambda product_id: OrderBook(), OrderBook.update)`

Reply:
I think that this was not yet included in the release, so I will note if bytewax version is greater than 0.9.0 to use what you wrote above.